File: stack/Valid Parentheses.py
def isValid(s: str) -> bool:
    """
    we run into a closing char that matches opening at top of stack we pop
    else we need to return false
    if it is an opening char then we add it to the stack
    """

    if len(s) <= 1:
        return False
    stack = []

    open = set(['(', '[', '{'])
    for c in s:
        if c in open:
            stack.append(c)
        else:
            if stack:
                top = stack.pop()
            else:
                return False
            if c == ')' and top != '(':
                return False
            elif c == ']' and top != '[':
                return False
            elif c == '}' and top != '{':
                return False

    return not stack

print(isValid("[]"))
print(isValid("([{}])"))
print(isValid("[(])"))
print(isValid("(("))

File: stack/Valid Parentheses.cs
public class SolutionIsValid
{
    public bool IsValid(string s)
    {
        if (s.Length == 1)
            return false;

        Stack<char> open = new Stack<char>();

        HashSet<char> opening = new HashSet<char> { '(', '[', '{' };

        for (int i = 0; i < s.Length; ++i)
        {
            char c = s[i];
            if (opening.Contains(c))
                open.Push(c);
            else
            {
                if (open.Count <= 0)
                    return false;

                char popped = open.Pop();
                if (c == ')' && popped != '(')
                    return false;
                if (c == ']' && popped != '[')
                    return false;
                if (c == '}' && popped != '{')
                    return false;
            }
        }
        return open.Count == 0;
    }
}
